src/jit/hwintrinsicxarch.cpp

```diff
@@ -754,6 +760,70 @@ GenTree* Compiler::impSSE2Intrinsic(NamedIntrinsic        intrinsic,
             }
             break;
 
+        case NI_SSE2_ConvertScalarToVector128Double:
+        {
+            assert(sig->numArgs == 2);
+            assert(getBaseTypeOfSIMDType(sig->retTypeSigClass) == TYP_DOUBLE);
+
+            argList = info.compCompHnd->getArgNext(sig->args);
+            CorInfoType corType =
+                strip(info.compCompHnd->getArgType(sig, argList, &argClass)); // type of the second argument
+
+            baseType = JITtype2varType(corType);
+
+#ifdef _TARGET_X86_
+            if (varTypeIsLong(JITtype2varType(corType)))
+            {
+                return impUnsupportedHWIntrinsic(CORINFO_HELP_THROW_PLATFORM_NOT_SUPPORTED, method, sig, mustExpand);
+            }
+#endif // _TARGET_X86_
+
+            if (baseType == TYP_STRUCT)
+            {
+                baseType = getBaseTypeOfSIMDType(argClass);
+                op2      = impSIMDPopStack(TYP_SIMD16);
+            }
+            else
+            {
+                op2 = impPopStack().val;
+            }
+
+            op1     = impSIMDPopStack(TYP_SIMD16);
+            retNode = gtNewSimdHWIntrinsicNode(TYP_SIMD16, op1, op2, intrinsic, baseType, simdSize);
+
+            break;
+        }
+
+        case NI_SSE2_ConvertScalarToVector128Single:
+        {
+            assert(sig->numArgs == 2);
+            assert(getBaseTypeOfSIMDType(sig->retTypeSigClass) == TYP_FLOAT);
+
+            argList = info.compCompHnd->getArgNext(sig->args);
+            CorInfoType corType =
+                strip(info.compCompHnd->getArgType(sig, argList, &argClass)); // type of the second argument
+
+            baseType = getBaseTypeOfSIMDType(argClass);
+
+            op2     = impSIMDPopStack(TYP_SIMD16);
+            op1     = impSIMDPopStack(TYP_SIMD16);
+            retNode = gtNewSimdHWIntrinsicNode(TYP_SIMD16, op1, op2, intrinsic, baseType, simdSize);
+            break;
+        }
+
+        case NI_SSE2_ConvertToInt32:
+        case NI_SSE2_ConvertToInt64:
+        case NI_SSE2_ConvertToUInt32:
+        case NI_SSE2_ConvertToUInt64:
+        {
+            assert(sig->numArgs == 1);
+            op1      = impSIMDPopStack(TYP_SIMD16);
+            retType  = JITtype2varType(sig->retType);
+            baseType = getBaseTypeOfSIMDType(info.compCompHnd->getArgClass(sig, sig->args));
```

Comment:
We can directly use return type as base type.

Reply:
There are overloads where it is necessary to select different instruction based on argument type - I use baseType to handle this. AFAIR during debugging I have seen problems with baseType detection and that's why I have explicitly recovered it from `sig->retType`.

Reply:
Ah, I see there are two special guys
```csharp
 /// <summary>
        /// int _mm_cvtsd_si32 (__m128d a)
        ///   CVTSD2SI r32, xmm/m64
        /// </summary>
        public static int ConvertToInt32(Vector128<double> value) => ConvertToInt32(value);
        /// <summary>
        /// int _mm_cvtsi128_si32 (__m128i a)
        ///   MOVD reg/m32, xmm
        /// </summary>
        public static int ConvertToInt32(Vector128<int> value) => ConvertToInt32(value);
        /// <summary>
        /// __int64 _mm_cvtsd_si64 (__m128d a)
        ///   CVTSD2SI r64, xmm/m64
        /// </summary>
        public static long ConvertToInt64(Vector128<double> value) => ConvertToInt64(value);
        /// <summary>
        /// __int64 _mm_cvtsi128_si64 (__m128i a)
        ///   MOVQ reg/m64, xmm
        /// </summary>
        public static long ConvertToInt64(Vector128<long> value) => ConvertToInt64(value);
```

Perhasps, we can separate these two groups
```cpp
case NI_SSE2_ConvertToInt32:
case NI_SSE2_ConvertToInt64:
    // get baseType from arg
case NI_SSE2_ConvertToUInt32:
case NI_SSE2_ConvertToUInt64:
    // use the return type as base type
```
because calling `getBaseTypeOfSIMDType` and `info.compCompHnd->getArgClass` are relatively expensive

Reply:
Yeah, you are right. Will fix